spmd/tensor/parallel/api.py

```diff
@@ -86,3 +91,93 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
+
+    Args:
+        module (nn.Module):
+            The :class:``nn.Module`` object to be parallelized.
+        device_mesh (DeviceMesh):
+            The n-d :class:``DeviceMesh`` over which the `module` will be parallelized.
+            If the mesh is more than 1-dimensional, we convert it to a 1-d :class:``DeviceMesh``
+            by `tp_mesh_dim`.
+        parallel_style (ParallelStyle):
+            :class:``ParallelStyle`` describes how weight and bias should be distributed
+            over :class:``DeviceMesh`` and how the input and output should be prepared for
+            Tensor Parallelism.
+            :class:``RowwiseStyle``: weight is sharded on dim 1 and bias is replicated.
+            :class:``ColwiseStyle``: weight and bias are both sharded on dim 0.
+            Default: :class:``ColwiseParallel``
+        tp_mesh_dim (int):
+            The dimension of :class:``DeviceMesh`` on which we perform
+            Tensor Parallelism.
+            Default: 0
+    Returns:
+        None
+    """
+
+    if not isinstance(module, nn.Linear):
+        raise RuntimeError(
+            f"_parallelize_linear expects a torch.nn.Linear module but received {module}!"
+        )
+
+    if not isinstance(parallel_style, ParallelStyle):
+        raise RuntimeError(
+            f"parallel_style passed to _parallelize_linear is not a ParallelStyle object but {type(parallel_style)}!"
+        )
+
+    col_wise_sharding: Sequence[Placement] = [Replicate()] * device_mesh.ndim
+    col_wise_sharding[tp_mesh_dim] = Shard(0)
+    row_wise_sharding: Sequence[Placement] = [Replicate()] * device_mesh.ndim
+    row_wise_sharding[tp_mesh_dim] = Shard(1)
+    replicate: Sequence[Placement] = [Replicate()] * device_mesh.ndim
+
+    def linear_module_placements(
+        parallel_style: ParallelStyle,
+    ) -> Tuple[Sequence[Placement], Sequence[Placement]]:
+        if isinstance(parallel_style, RowwiseParallel):
+            return row_wise_sharding, replicate
+        elif isinstance(parallel_style, ColwiseParallel):
+            return col_wise_sharding, col_wise_sharding
+        else:
+            raise RuntimeError(f"{type(parallel_style)} is not supported!")
+
+    # placements
+    weight_placements, bias_placements = linear_module_placements(
+        parallel_style
+    )
+    # params
+    weight = nn.Parameter(
+        distribute_tensor(module.weight, device_mesh, weight_placements)
+    )
+    module.register_parameter("weight", weight)
+    if module.bias is not None:
+        bias = nn.Parameter(
+            distribute_tensor(module.bias, device_mesh, bias_placements)
+        )
+        module.register_parameter("bias", bias)
+    # input
+    if parallel_style._prepare_input is not None:
+        input_fn: Callable[
+            [Sequence[Union[torch.Tensor, DTensor]], Tuple[DTensor, ...]]
+        ] = lambda inputs: tuple(
+            parallel_style._prepare_input(tensor, device_mesh)
+            for tensor in inputs
```

Comment:
I think we can just focus on inputs[0] for now and we can use `distribute_module` API.

Reply:
Will address it later.